Emit fix-its for `Return from initializer` missing variables

iMostfa opened the pull request:
Resolves https://github.com/swiftlang/swift/issues/80944



https://github.com/user-attachments/assets/9b0bd10f-4457-4610-bd30-1341a2cf38ca



This's my proposed solution for the issue.

the proposed solution: 

1. Check the currently passed parameters in the `init(....)`  function
2. suggest a fix it on the `init` to add `,foo: Bar` 
3. suggest a fix on the closing brace, to add `self.foo = foo after {` which is the opening of brace of init function

take the as input: 

```swift
struct Foo {
    let name: String
    let k: (String, Int)
    let m: String

    init(name: String, m: String) {
        self.name = name
    }
}

```

the following fix-its are emitted: 

```json
[
 {
  "file": "/Users/mostafa.essam/Desktop/file2.swift",
  "offset": 143,
  "text": "self.k = k\nself.m = m\n"
 },
 {
  "file": "/Users/mostafa.essam/Desktop/file2.swift",
  "offset": 110,
  "text": ",k: (String, Int)"
 }
]

```


when applied, it will result to: 

```swift
struct Foo {
    let name: String
    let k: (String, Int)
    let m: String

    init(name: String, m: String, k: (String, Int)) {
        self.name = name
        self.k = k
        self.m = m
    }
}
```


i opened the PR to discuss the solution

iMostfa committed:
Emit Fixit for missing init varaibles

Avoid string dances

iMostfa commented:
Hi @eeckstein would love to have your input here 🙏🏻

eeckstein commented:
I'm probably not the best person to review this. Maybe @rintaro can take a look?
But please first bring your changes in a better shape. For example, indentation is completely wrong.
Also, I don't understand your motivating example in the description.

iMostfa commented:
@eeckstein I updated the example; I had accidentally pasted the output into the input earlier 🫠
i also updated the indentation. 

additionally, i update the PR with a video showing the change from Xcode, a little bit easier when visualized.

iMostfa commented:
@rintaro would love to have your input on my PR please 🙏🏻

iMostfa commented:
@eeckstein is it possible to assign @rintaro as a PR Reviewer ?

rintaro commented:
I can review the diagnostic code, but I'm not familiar with these DI APIs e.g. `DIMemoryUse`. From the history maybe @slavapestov ?

But before that, please add some test cases. You could update `test/SILOptimizer/definite_init_*.swift`

iMostfa committed:
Fix Swift tests

iMostfa committed:
Format code

iMostfa commented:
tests updated with the new expected fix-its. 
can you @slavapestov take a look ? 👀

iMostfa commented:
ping 🙏🏻

eeckstein commented:
cc @kavon for the DI changes and @rintaro (again)
